### DISCUSSION TITLE
Python language projections

### DISCUSSION THREAD

**Discussion by LeonarddeR** [microsoft/win32metadata#374] [Ideas]
Hereby I'd like to suggest the creation of Python language projections for the Win32 API. I'm one of the developers of the [NVDA screen reader](https://nvaccess.org) that is written in Python for the most part and heavily relies on talking to the Win32 API. Currently this is performed using ctypes.

**Comment by kennykerr** [comment]
See https://github.com/microsoft/win32metadata/issues/174#issuecomment-766284319

**Comment by mikebattista** [comment]
#183 captures the guidance for new language projection requests.

**Comment by mikebattista** [comment]
@zooba and I met to brainstorm what an ideal Python projection would look like based on learnings from projects like pywin32 and ctypes. Below are the notes from that discussion for review and comment.

* Provide maximum API coverage for Win32 and COM
* APIs should be exposed as friendly Python interfaces rather than like ctypes
* Achieve seamless acquisition and “pay for play” by providing granular PIP packages under the [windows ](https://pypi.org/project/windows/) root
  * These packages would probably align with the namespaces we’ve added in the Win32 metadata
  * Ideally, the tooling used to build these packages could be used by developers directly to produce their own packages for just the APIs they need, if the Win32 namespaces aren’t granular enough
* These packages sim-ship with Python updates so they are available at launch

Let us know your thoughts and comments.

**Reply by LeonarddeR** [reply]
> @zooba and I met to brainstorm what an ideal Python projection would look like based on learnings from projects like pywin32 and ctypes. Below are the notes from that discussion for review and comment.

Very interesting!

How would this projection call the real underlying API? I see several possibilities:

* Friendly Python functions that call the underlying API using ctypes. Probably have to work around incompleties in ctypes.
* Using Python C(++) extensions
* Using cffi

Also, how about [PyWinRT](https://github.com/Microsoft/xlang/tree/master/src/package/pywinrt/projection)?

**Reply by mikebattista** [reply]
From the discussion, it sounded like there were performance concerns with ctypes, so we probably wouldn't build on that, especially if you're saying it may not provide all the support we'd need.

I believe native extensions was the idea, though I'm not sure the pros and cons of that vs. cffi. Is there one approach you feel would be superior?

Regarding Python/WinRT, we would think holistically about the entire Windows API surface as part of this. The [windows](https://github.com/microsoft/windows-rs) crate for Rust supports all Windows APIs and achieves "pay for play" via source generation. If a similar model is not possible here, would we want to apply the namespace-based package approach for WinRT as well rather than one package for all of WinRT?

**Reply by LeonarddeR** [reply]
To be honest, I have no personal experience with cffi myself. It means an extra dependency that's not part of Python anyway. Since PyWinRT goes the native extension way.

> If a similar model is not possible here, would we want to apply the namespace-based package approach for WinRT as well rather than one package for all of WinRT?

I don't have a strong opinion about this, but I"m not sure about the namespace based approach actually. There are certainly advantages  as long as there's a meta package that bundles them all.

Isn't PyWinRT based on C++/WinRT actually? In that case, it might be better to see how plans will go to bundle C++/WinRT and C++/Win32 in one package. See https://github.com/microsoft/cppwin32/issues/2

**Reply by sander76** [reply]
I would very, very much welcome better windows support for python. My current interests would go out to bluetooth support and being able to properly access a webcam. Both I am currently trying to do with winrt.

As a sidequestion: how would this relate to winrt ?

**Reply by mikebattista** [reply]
As I mentioned earlier, we would think holistically about all Windows APIs. We currently have an experimental [Python/WinRT](https://github.com/Microsoft/xlang/tree/master/src/package/pywinrt/projection) projection, so we'd need to think about finishing that and how we could potentially unify the tools and developer experience for all Windows APIs like we did with the [windows](https://github.com/microsoft/windows-rs) crate for Rust.

That said, the specific work proposed here is for a Python projection for Win32 and COM.

**Reply by LeonarddeR** [reply]
> That said, the specific work proposed here is for a Python projection for ... COM.

Does that mean you're intending to provide a full blown alternative to comtypes? Or only for the WIN32 stuff that'relies on it?

**Reply by mikebattista** [reply]
We are looking at existing projects in the space as part of this. Obviously we're not going to build something if everyone's happy with existing support, though what we typically hear across languages is that manually-maintained projects that provide friendly interfaces over Win32 and COM APIs have spotty coverage and can be slow moving with high maintenance overhead for the community. A metadata-based projection would help with that.

**Reply by lunixbochs** [reply]
> manually-maintained projects that provide friendly interfaces over Win32 and COM APIs have spotty coverage and can be slow moving with high maintenance overhead for the community

This is basically how I'd describe the current Windows API libraries for Python. I would strongly prefer a proper win32/com projection for Python instead.

My vote would be to avoid CFFI/ctypes, and go straight for an extension module with precompiled packages for any non-EOL cpython versions. Long term can migrate the extension to whatever becomes of [HPy](https://github.com/hpyproject/hpy) to provide better support for GC'd Python interpreters like pypy.

(I ended up in this thread after looking at the comtypes client for an API and hoping there was a projection I could use instead. Alas.)

**Reply by LeonarddeR** [reply]
I like this idea.

It is probably best to do something based on [C++ Win32](https://github.com/microsoft/cppwin32). Aren't the [WinRT bindings for Python](https://pypi.org/project/winrt/) based on C++/WinRT as well?

**Reply by lunixbochs** [reply]
Something else to watch out for here is: I think bundling a whole namespace together into one Python package means you're forced to the max Windows version and runtime libraries needed for the whole namespace. In `windows-rs`, I noticed that binding `*` from a namespace would incur additional runtime library dependencies.

One way to work around that could be to split a namespace into multiple numbered packages that indicate the minimum Windows version?

**Comment by byehack** [comment]
I just should mention https://github.com/ynkdir/py-win32more, the win32metadata based python package.
it's not stable after few months (api may changed in future releases) but still provides good usability.

**Reply by mikebattista** [reply]
Awesome! I've added it to the list of projections in the README.

**Reply by LeonarddeR** [reply]
The readme of this project mentions both unmaintained and untested, so I'm afraid that's a no go for the distant future.

**Comment by LeonarddeR** [comment]
I've been pondering for several days about something @kennykerr can probably assess for feasibility.
There is the [PyO3](https://pyo3.rs/) Rust crate that allows creating Python bindings from Rust. A custom project could create wrappers around the windows or windows-sys crate, but it is probably more effectieve to adapt the binding generator to create PyO3 compatible structs, enums and functions to begin with.
